graphdatascience/query_runner/protocol/project_protocols.py

```diff
@@ -86,3 +92,41 @@ def run_projection(
     ) -> DataFrame:
         versioned_endpoint = ProtocolVersion.V2.versioned_procedure_name(endpoint)
         return query_runner.call_procedure(versioned_endpoint, params, yields, database, logging, False)
+
+
+class ProjectProtocolV3(ProjectProtocol):
+    def project_params(
+        self, graph_name: str, query: str, job_id: str, params: Dict[str, Any], arrow_config: Dict[str, Any]
+    ) -> CallParameters:
+        return CallParameters(
+            graph_name=graph_name,
+            query=query,
+            job_id=job_id,
+            arrow_configuration=arrow_config,
+            configuration={
+                "concurrency": params["concurrency"],
+                "undirectedRelationshipTypes": params["undirected_relationship_types"],
+                "inverseIndexedRelationshipTypes": params["inverse_indexed_relationship_types"],
+            },
+        )
+
+    def run_projection(
+        self,
+        query_runner: QueryRunner,
+        endpoint: str,
+        params: CallParameters,
+        yields: Optional[List[str]] = None,
+        database: Optional[str] = None,
+        logging: bool = False,
+    ) -> DataFrame:
+        def is_not_done(result: DataFrame) -> bool:
+            result = project_fn().squeeze()
+            return result["status"] != Status.DONE.name
+
+        @retry(retry=retry_if_result(is_not_done), wait=wait_incrementing(start=0.2, increment=0.2, max=2))
+        def project_fn() -> DataFrame:
+            return query_runner.call_procedure(
+                ProtocolVersion.V3.versioned_procedure_name(endpoint), params, yields, database, logging, False
+            )
```

Comment:
I did not follow this work to closely.
How do we avoid that the client will not send the previous batches again?

Or is it rather, that the Arrow client will continue with the projection even on a connection loss?
And we will run the cypher part up to the aggregation function, which find an existing projection and returns early?
Ideally we could check upfront if a projection with the job-id already exists?

Reply:
The arrow part is unrelated to this change and will be implemented in another PR (we didn't implement it yet on the dbms or session side). What we did here is to guard the procedure execution from connection losses. Before, when the bolt connection to a project/write procedure was interrupted, the transaction of that procedure was killed and therefore the arrow connection, leading to a complete interruption of the work.
Now, the procedures will spawn tasks that run in the background instead, independent of transactions. When calling a procedure one of the following 2 things will happen:
1. If there is no process with the given job-id already running, spawn a process and return
2. If a process is running, check if it is terminated and return an appropriate status

Here in the python client we merely have to implement the status polling, everything else is handled within the procedure implementations.